### PR TITLE
[FLINK-20846] Decouple checkpoint services and CheckpointCoordinator

### PR DESCRIPTION
**Pull request by tillrohrmann**
In order to support reusing checkpoints across different `ExecutionGraph` instances, this PR decouples the checkpoint services `CompletedCheckpointStore` and `CheckpointIDCounter` from the lifetime of the `CheckpointCoordinator`. Concretely, this means that these services won't be shut down by the `CheckpointCoordinator` but by an instance which can live longer than the `CheckpointCoordinator`. 

The current owner of the checkpoint services is the `SchedulerBase`. The `SchedulerBase` shuts the services down once the `executionGraph` reaches a terminal state.

* 7e2af0e Factor out CompletedCheckpointStore and CheckpointIDCounter creation in ExecutionGraphBuilder.buildGraph
* ab4b146 Move CheckpointIDCounter creation out of ExecutionGraphBuilder
* c5b6e2c Move CompletedCheckpointStore creation out of ExecutionGraphBuilder.buildGraph
* 040cffc Remove CheckpointRecoveryFactory from ExecutionGraphBuilder.buildGraph
* fa3309b Move checkpoint service shut down out of CheckpointCoordinator

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphDeploymentTest.java [hunk 1]
```diff
@@ -742,25 +731,11 @@ private ExecutionGraph createExecutionGraph(Configuration configuration) throws
...
                         null));
 
-        final Time timeout = Time.seconds(10L);
-        return ExecutionGraphBuilder.buildGraph(
-                null,
-                jobGraph,
-                configuration,
-                executor,
-                executor,
-                new ProgrammedSlotProvider(1),
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphDeploymentTest.java:751]: Seems like this was the only remaining usage of the `ProgrammedSlotProvider`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerUtils.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                        jobId,
+                        "Failed to initialize high-availability completed checkpoint store",
+                        e);
+            }
+        } else {
+            return DeactivatedCheckpointCompletedCheckpointStore.INSTANCE;
+        }
+    }
+
+    static CompletedCheckpointStore createCompletedCheckpointStore(
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerUtils.java:66]: `@VisibleForTesting`?

**Review by zentol — APPROVED** [review]

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphDeploymentTest.java (hunk 1 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphDeploymentTest.java:751]: Oh yes, good catch. I'll remove it :-)

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerUtils.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerUtils.java:66]: Yes, you are right. I'll update it.

**Comment by tillrohrmann** [comment]
Thanks for the review @zentol. I've addressed your comments and pushed an update. Merging once AZP gives green light.

**Pull request closed without merging**